snapshot/git/gitdb/db.go

```diff
@@ -0,0 +1,155 @@
+package gitdb
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/scootdev/scoot/os/temp"
+	"github.com/scootdev/scoot/snapshot"
+	"github.com/scootdev/scoot/snapshot/git/repo"
+)
+
+// valueKind describes the kind of a Value: is it a Snapshot or a SnapshotWithHistory?
+// kind instead of type because type is a keyword
+type valueKind int
+
+const (
+	kindSnapshot valueKind = iota
+	kindSnapshotWithHistory
+)
+
+// MakeDB makes a gitdb.DB that uses dataRepo for data and tmp for temporary directories
+func MakeDB(dataRepo *repo.Repository, tmp *temp.TempDir) *DB {
+	return &DB{
+		dataRepo: dataRepo,
+		tmp:      tmp,
+	}
+}
+
+// DB stores its data in a Git Repo
+type DB struct {
+	dataRepo *repo.Repository
+	tmp      *temp.TempDir
+}
+
+// IngestDir ingests a directory directly.
+// The created value is a Snapshot.
+func (db *DB) IngestDir(dir string) (snapshot.ID, error) {
+	// We ingest a dir using git commands:
+	// First, create a new index file.
+	// Second, add all the files in the work tree.
+	// Third, write the tree.
+	// This doesn't create a commit, or otherwise mess with repo state.
+	indexDir, err := db.tmp.TempDir("git-index")
+	if err != nil {
+		return "", err
+	}
+
+	indexFilename := filepath.Join(indexDir.Dir, "index")
+	defer os.RemoveAll(indexDir.Dir)
+
+	extraEnv := []string{"GIT_INDEX_FILE=" + indexFilename, "GIT_WORK_TREE=" + dir}
+
+	// TODO(dbentley): should we use update-index instead of add? Maybe add looks at repo state
+	// (e.g., HEAD) and we should just use the lower-level plumbing command?
+	cmd := db.dataRepo.Command("add", ".")
+	cmd.Env = append(cmd.Env, extraEnv...)
+	_, err = db.dataRepo.RunCmd(cmd)
+	if err != nil {
+		return "", err
+	}
+
+	cmd = db.dataRepo.Command("write-tree")
+	cmd.Env = append(cmd.Env, extraEnv...)
+	sha, err := db.dataRepo.RunCmd(cmd)
+	if err != nil {
+		return "", err
+	}
+	sha, err = repo.ValidateSha(sha)
+	if err != nil {
+		return "", err
+	}
+
+	v := &localValue{sha: sha, kind: kindSnapshot}
+	return v.ID(), nil
+}
+
+// IngestGitCommit ingests the commit identified by commitish from ingestRepo
```

Comment:
I haven't been duplicating interface comments in my impls. Do we want to duplicate or breadcrumb the interface comments?